contributors/design-proposals/storage/local-storage-pv.md

```diff
@@ -761,3 +761,202 @@ The central manager will poll each worker for their discovered volumes and creat
+
+* list/get PersistentVolumeClaims
+* list/get/update StorageClasses
+
+##### Volume Creation
+
+During start-up of the provisioner, a new channel `provisionChan` will be created,
+and passed into both `populator` and `storageManager`:
+
+```
+provisionChan chan *v1.PersistentVolumeClaim
+```
+When it finds a PVC is in need of privsioning, as desicribed above,
+the `populator` will pass it into the channel to trigger volume privisioning;
+`StorageManager` will be the consumer, it will watch the channel and
+manage provisioning when new input received, that is:
+
+1. Create volumes behind a PV accordingly
+2. Create a PV object that is pre-bound to the PVC
+
+For LVM, the proposed workflow is:
+
+1. Check if there is already a PV pre-bound to the PVC, if so, return directly
+(This can happen when a volume has already been provisioned for a PVC,
+but has not been bound to the PVC by the controller yet).
+2. Check if Storage Class of the PVC is in record of its storageClass map.
+If not, return error.
+3. Check if the class is available for dynamic provisioning (i.e. has an input of `volumeGroup`).
+If not, return error.
+4. Check if `Free Size` of the VG under the class can satisfy the capacity requested
+in the PVC. If not, return error.
+5. Create `logical volume (LV)` object in the VG.
+6. Create a PV API object that is pre-bound to the PVC, and the path
+of local storage in the PV should be the path of the LV created above.
+
+##### Cleanup after Release
+
+Currently we already have a component `deleter` to cleanup volumes/PVs of static local volume,
+detail cleanup actions are managed via custom scripts/binaries that specified in `blockCleanerCommand`.
+
+Existing script `quick_reset.sh` calls `mkfs -F` and `wipefs` to cleanup a block.
+In addition to that, a hook named `volumeDeleteFunc` will be added into `deleter`,
+and for dynamic privisioned volumes, the deleting process should be:
+1. Call scripts/binaries in `blockCleanerCommand` to clean up data in the volume.
+2. Call `volumeDeleteFunc` to recycle the volume behind the PV (e.g. delete the LV for lvm).
+3. Delete the PV object.
+
+##### Expansion of LocalVolumeSource
+
+Currently, the path in `LocalVolumeSource` must be a directory. Thus, for dynamic
+provisioned volumes, the provisioner need to mount the LV to a host path first,
+to allow kubelet for another mount. It increases the complexity of the plugin.
+
+Therefore, in-tree local storage plugin will be updated to support both directory
+and block as a source of mount. The mount process would be:
+1. Detect if given path in volumeSource is mounted or not, if mounted,
+continue to work in the existing way.
+2. If not, Check if global path `{pluginDir}/mounts/{volumeID}` exists.
+and create the directory if not.
+3. Check if the global path is of mount point, if not, mount given path to
+the global path.
+4. Manage `mkfs` for the global path if it is newly created or mounted.

```

Comment:
How will you handle multiple pods trying to mkfs and/or mount the same volume at the same time?

Reply:
I think there were some updates made to the Attacher interface to better support local attaching volume types, that should only be called once per node, instead of once per pod.  I wonder if it could be used here.  @saad-ali

Reply:
Oh, then the global path tasks should indeed be managed in Attacher/Detacher. Otherwise, it's hard to solve the race condition, and hard to decide when to unmount the global path.

Reply:
Walking through the code, I think the volumeManager process the mounts in order, seems it will not meet a race condition?

Anyway, refer to the behavior of other plugins, it seems reasonable to implement interface `MountDevice`/`UnmountDevice` of Attacher/Detacher to manage the mount of global path.
The downside is we haven't seen real requirement on attach for local storage, so interface like  `Attach`/`VolumesAreAttached`/`WaitForAttach` will be left empty.

By the way, the new `MountDevice` is still in the process of `MountVolume`, @msau42 , can you see race condition this way?

Reply:
Another thing to note is that other attachable storages will have a global path for sure, and unmount it in `UnmountDevice` without check.
But for local storage, if given path is a already directory, the global path is not needed, thus in `UnmountDevice`, we might need a basis to check whether to do unmount

Reply:
I spoke with @saad-ali and @jingxu97.  If the plugin does not implement the Attach interface, then it is possible for Mount/Unmount to be called concurrently across pods sharing the same volume, so I think we'll have to implement an Attach interface to do the global mount.  However, there is a problem that the Attach call is run on the master, and not on the node.  So some plugins, like iscsi and fc, have implemented their "local attach" in the WaitForAttach call, which is not the cleanest solution.  @saad-ali prefers that we implement a new LocalAttach interface, and eventually we can also migrate other plugins to this interface too.  Some more discussion on local attach is [here](https://github.com/kubernetes/kubernetes/issues/54108)

cc @mlmhl

Reply:
Yeah, I'm working on this, according doc and PR is #1683  and kubernetes/kubernetes#58810. We add a new `DeviceMounter` interface to do the local attach operation.

However, if we want the local attach operation called once per node, maybe we need to do more work: Currently the `WaitForAttach`, `MountDevice` and `SetUp` operations are all performed inside `operationGenerator.GenerateMountVolumeFunc`, which is called once per pod. One possible solution is we split the `WaitForAttach` and `MountDevice` operation into a single function(for example: `GenerateMountDeviceFunc`) which called once a node(i.e. only called if the volume isn't globally mounted).
@saad-ali  @msau42  Any thoughts?

Reply:
If volume manager can coordinate the serialization of MountDevice and MountVolume, then I think getting rid of WaitForAttach is fine.

Reply:
Discussed with @mlmhl, [the refactor](https://github.com/kubernetes/kubernetes/pull/58810) does not cover the change in the process of `operationGenerator`, maybe we can raise another issue to address the race scenario?

Reply:
It's indeed a common case, but a new interface seems duplicated to exsiting `MountDevice` interface. I think we can reuse the  interface to manage local attach. It's the fact that the interface included in `Attacher` that confusing us.
That is also what https://github.com/kubernetes/kubernetes/pull/58810 intend to address.

After that, we need a further step to make funcs in `DeviceMounter` called per node I think.
@msau42 @mlmhl WDYT?

Reply:
I discussed this with @jingxu97.  We think the best approach would be to split this into 2 phases:

1. Separate Attach calls from MountDevice.  Handled by #1683 
2. Move Attach + MountDevice to outside the MountVolume operation.  This is a much bigger change and would impact all volume plugins, so needs to be very carefully considered.

So for now, we can implement MountDevice.  It's going to get called per pod, however, we must make sure operation executor will serialize the MountVolume operation per volume (right now it's only done for attachable plugins).

Reply:
sounds good to me.
>  It's going to get called per pod, however, we must make sure operation executor will serialize the MountVolume operation per volume (right now it's only done for attachable plugins).

Yah, maybe improve `pendingOperations` in operation executor to also keeps track of mount operations.
Anyway, I think we should open another  issue to address it. and what we should describe here is after #1683 in, how we should implement related interfaces in local volume plugin to support both block and file path as volume source.